grl/policy_eval.py

```diff
@@ -36,11 +39,13 @@ def solve_mdp(self, mdp):
             a_t = np.zeros(mdp.n_states)
             a_t[s] = -1 # subtract V_pi(s) to right side
             b_t = 0
-            possible_next_ss = np.where(mdp.T[self.pi[s],s] != 0.)[0]
-            for next_s in possible_next_ss:
-                t = mdp.T[self.pi[s],s,next_s]
-                b_t -= t * mdp.R[self.pi[s],s,next_s] # subtract constants to left side
-                a_t[next_s] += t * mdp.gamma
+            t = mdp.T * self.pi_ground[s][:, None, None]
+            r = mdp.R * self.pi_ground[s][:, None, None]
+            for next_s in range(mdp.n_states):
+                t_n = t.sum(0)[s,next_s]
+                r_n = r.sum(0)[s,next_s]
```

Comment:
Rather than broadcasting and then summing along the actions dimension, we can probably just ~matrix multiply~ take the tensor dot product.

If `T` is `|A| x |S| x |S|` and `pi[s]` is length `|A|` then `np.tensordot(pi[s], T, axes=1)` is `|S| x |S|`.

I would call the result `T_pi`.

Reply:
Nice, done